### PR TITLE
blog: A new ProtoBuf generator for Go

### PR DESCRIPTION
**Pull request by vmg**
Here's the first iteration of this blog post.

cc @systay @deepthi @shlomi-noach for review

### CODE REVIEW

**Comment by vmg** [comment]
Rendered: https://github.com/vitessio/website/blob/5acae49a28ad225424bf8a3434a2695e860e663d/content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md

**Comment by shlomi-noach** [comment]
Real rendered: https://deploy-preview-770--vitess.netlify.app/blog/2021-06-03-a-new-protobuf-generator-for-go/?utm_source=github&utm_campaign=bot_dp

**Review by systay — APPROVED** [review]

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+Although the new (un)marshaling code in ProtoBuf APIv2 is _not_ measurably slower than the one in APIv1 (it is, in fact, mostly equivalent), Vitess hasn't been using the APIv1 codecs for a while. Earlier this year, we introduced the Gogo ProtoBuf compiler to our codebase, with really impressive performance results.
+
+For those who are not aware of it, [Gogo ProtoBuf](https://github.com/gogo/protobuf) is a _fork_ of the original ProtoBuf APIv1 that includes a custom code generator with optional support for many performance related features. The most notable of them, and the one we enabled for Vitess', is the generation of fully unrolled marshaling and unmarshaling code for all the messages in the codebase. This is a very significant performance boost compared to marshaling using the default codecs in ProtoBuf APIv1, which are fully implemented using reflection at runtime. Having static code that can be compiled ahead of time results in much lower CPU usage and measurably shorter response times for our most demanding RPC calls.
```
Comment (deepthi) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: typo: `the one we enabled for Vitess'` -> `the one we enabled for Vitess` (remove unnecessary apostrophe)

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+Hence, it was clear to us that introducing Gogo ProtoBuf generated code in Vitess would be a very short-lived optimization: as soon as we decided to upgrade our project to ProtoBuf APIv2, we'd have to drop Gogo altogether and go back to the default reflection-based (un)marshaler, but we still took the plunge, making sure to use as few Gogo-exclusive features as possible to make the eventual upgrade less painful.
+
+Was this a bad idea? Maybe. Earlier this month we decided to attempt upgrading Vitess to ProtoBuf APIv2, mostly to find out how hard would the process be, and how large the performance regression when removing Gogo's autogenerated code. The upgrade didn't go without its [hiccups](https://github.com/vitessio/vitess/pull/8075), but we managed to get Vitess working with the new API and the test suite fully green after a couple weeks of effort.
```
Comment (deepthi) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: couple weeks -> couple of weeks

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+The results of our benchmarks, however, were discouraging. [AreWeFastYet](https://benchmark.vitess.io/), our nightly benchmark system, detected a very significant increase in CPU usage during all benchmarking runs: up to 19%, resulting in lowering the total throughput of of the system by roughly 3%.
+
+![](https://imgur.com/7Of96gT.png)
+
+We have a very stringent rule of never regressing performance between Vitess releases, so we started evaluating our options to perform this upgrade and leaving behind the ProtoBuf APIv1 package (now deprecated) and the Gogo ProtoBuf compiler (now unmaintained) while keeping Vitess as fast as it was.
+
+## Introducing `vtprotobuf`, the Vitess ProtoBuf compiler
+
+The most obvious choice that occured to us was picking up ownership of Gogo ProtoBuf, but its maintainers were right: upgrading it to APIv2 is a massive undertaking. We don't have the time or manpower to commit to upgrading _and_ owning such project indefinitely.
```
Comment (deepthi) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: humanpower :grinning: 
jk, you can leave it in or re-word as you wish.

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+Hence, the next best option are memory pools for individual objects, and this is the functionality that `vtprotobuf` now provides. When enabling the `pool` feature in `protoc-gen-go-vtproto`, we can mark individual ProtoBuf messages so they're always allocated off a memory pool when being unmarshaled. The marked messages receive a helper method so they can be directly returned to the memory pool when they're no longer needed, and a specialized `Reset` implementation that zeros them out whilst making sure to keep as much underlying memory around as possible (e.g. in nested slices, maps and objects) so that retrieving those objects from the memory pool later on results in less allocations when unmarshaling.
```
Comment (deepthi) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: nit: less allocations -> fewer allocations

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+Memory pooling is far from a performance silver bullet: many ProtoBuf messages are too small to benefit from pooling at all, and some have so much nested complexity that the overhead of recursively pooling its fields defeats the optimization in the first place. Furthermore, the Go programming language provides no affordances whatsoever in its type system to safely manage these pooled objects. If not used very carefully, it's very easy to corrupt memory and cause data races with pooled ProtoBuf messages.
```
Comment (deepthi) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: I'm not sure exactly what affordances means here. Do you mean the same thing as guarantees?

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+To measure the impact of ProtoBuf changes in our replication process, we've built a synthetic stress test that performs a full replication of a large MySQL table between two Vitess namespaces. This is a realistic example that exercises the most expensive part of the replication process: the initial copy of all row data between the two Vitess nodes.
+
+As a baseline for our our performance measurements, we're going to perform the replication using an old Vitess commit, where we were still using ProtoBuf APIv1 but without any of the Gogo ProtoBuf optimizations.
```
Comment (deepthi) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: typo: `our our` -> `our`

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+## Case study: VReplication Row Streaming
+
+Vitess' replication engine, VReplication, is one of the most RPC-heavy paths in a Vitess cluster. It is a powerful and versatile MySQL replication engine, capable of keeping copies of tables in sync between distinct Vitess namespaces and materializing partial views of tables. Whenever we commit changes or optimizations that affect GRPC or ProtoBuf marshaling, we pay very close attention to their impact on the replication subsystem. Since the whole replication process is performed through GRPC, and in production deployments the replicated tables are often massive (terabytes), any performance regressions in this path often result in real world time increases that range between minutes and hours.
+
+To measure the impact of ProtoBuf changes in our replication process, we've built a synthetic stress test that performs a full replication of a large MySQL table between two Vitess namespaces. This is a realistic example that exercises the most expensive part of the replication process: the initial copy of all row data between the two Vitess nodes.
```
Comment (deepthi) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: Might be better to use keyspace / tablet terminology
namespaces -> keyspace
node -> tablet

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+Once the pooling helpers for our messages have been generated, we must update the calling code to ensure we're pooling the messages from the stream. For this specific `VStreamRows` API, Vitess already provides a callback-based abstraction behind the stream, so fetching and returning the messages from the pool is very easy to implement, assuming that receivers of the `send(r)` callback do not keep the message around -- which they don't.
+
+![](https://imgur.com/Nl48Ulp.png)
+
+With these trivial changes in place, we can run our replication benchmark again and look at the overhead of unmarshaling the rows:
+
+![](https://imgur.com/p9v33OD.png)
+
+No, I did not screw up when cropping the flame graph. Once we enable memory pooling for `VStreamRowsResponse` objects, the unmarshaling code no longer needs to allocate memory for the underlying `Row` data. We're spending **0.63 CPU seconds** in unmarshaling now, because we just keep re-using the same few `Response` objects over and over again while copying rows from the source Vitess node.
```
Comment (deepthi) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: source Vitess node -> source Vitess tablet (??)

**Review by deepthi — APPROVED** [review]
Beautifully written.
I have a few nits/suggestions, but otherwise this looks great!

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md (hunk 3 above)
Comment (shlomi-noach) [reply content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: better reword it

**Comment by vmg** [comment]
Thanks for the review @deepthi! Very helpful! Updated the post.

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+Was this a bad idea? Maybe. Earlier this month we decided to attempt upgrading Vitess to ProtoBuf APIv2, mostly to find out how hard would the process be, and how large the performance regression when removing Gogo's autogenerated code. The upgrade didn't go without its [hiccups](https://github.com/vitessio/vitess/pull/8075), but we managed to get Vitess working with the new API and the test suite fully green after a couple of weeks of effort.
+
+The results of our benchmarks, however, were discouraging. [AreWeFastYet](https://benchmark.vitess.io/), our nightly benchmark system, detected a very significant increase in CPU usage during all benchmarking runs: up to 19%, resulting in lowering the total throughput of of the system by roughly 3%.
+
+![](https://imgur.com/7Of96gT.png)
+
+We have a very stringent rule of never regressing performance between Vitess releases, so we started evaluating our options to perform this upgrade and leaving behind the ProtoBuf APIv1 package (now deprecated) and the Gogo ProtoBuf compiler (now unmaintained) while keeping Vitess as fast as it was.
```
Comment (shlomi-noach) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: > a very stringent rule

Sounds like a lot of commitment. I'm sure some users will point out some regressions that took place. Let's reword towards something like "we strongly prefer" or similar.

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+Memory pooling is far from a performance silver bullet: many ProtoBuf messages are too small to benefit from pooling at all, and some have so much nested complexity that the overhead of recursively pooling its fields defeats the optimization in the first place. Furthermore, the Go programming language provides no capabilities in its type system to safely manage these pooled objects. If not used very carefully, it's very easy to corrupt memory and cause data races with pooled ProtoBuf messages.
```
Comment (shlomi-noach) [inline content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md:57]: I really appreciate the disclaimers.

**Review by shlomi-noach — APPROVED** [review]
Loved it!

File: content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md (hunk 9 above)
Comment (vmg) [reply content/en/blog/2021-06-03-a-new-protobuf-generator-for-go.md]: Good point :+1:

**Comment by shlomi-noach** [comment]
Going out today?

**Comment by vmg** [comment]
Yeah! I'll post at around 6:30PM my time, that's usually the most successful time we've seen for Twitter/HN/Reddit/etc.

**Comment by fsaintjacques** [comment]
Out of extreme timing and luck, I was *just* googling `gogo golang protobuf arena` to see if there was any prior work done on the subject, and low and behold I stumbled on the post (props to google for the fast indexing!?).

**Pull request merged**